lib/oban/postgres_notifier.ex

```diff
@@ -0,0 +1,201 @@
+defmodule Oban.PostgresNotifier do
+  @doc """
+  Postgres Listen/Notify based Notifier
+  """
+
+  @behaviour Oban.Notifier
+
+  use GenServer
+
+  import Oban.Breaker, only: [open_circuit: 1, trip_circuit: 3]
+
+  alias Oban.{Config, Registry, Repo}
+  alias Postgrex.Notifications
+
+  @mappings %{
+    gossip: "oban_gossip",
+    insert: "oban_insert",
+    signal: "oban_signal"
+  }
+
+  @channels Map.keys(@mappings)
+
+  defmodule State do
+    @moduledoc false
+
+    @enforce_keys [:conf]
+    defstruct [
+      :conf,
+      :conn,
+      :name,
+      circuit: :enabled,
+      reset_timer: nil,
+      listeners: %{}
+    ]
+  end
+
+  defguardp is_channel(channel) when channel in @channels
+
+  @impl Oban.Notifier
+  def start_link(opts) do
+    name = Keyword.get(opts, :name, __MODULE__)
+
+    GenServer.start_link(__MODULE__, Map.new(opts), name: name)
+  end
+
+  @impl Oban.Notifier
+  def listen(server \\ Oban, channels)
+
+  def listen(pid, [_ | _] = channels) when is_pid(pid) do
+    :ok = validate_channels!(channels)
+
+    GenServer.call(pid, {:listen, channels})
+  end
+
+  def listen(oban_name, [_ | _] = channels) do
+    oban_name
+    |> Registry.whereis(Oban.Notifier)
+    |> listen(channels)
+  end
+
+  @impl Oban.Notifier
+  def unlisten(oban_name \\ Oban, [_ | _] = channels) do
+    oban_name
+    |> Registry.whereis(Oban.Notifier)
+    |> GenServer.call({:unlisten, channels})
+  end
+
+  @impl Oban.Notifier
+  def notify(%Config{} = conf, channel, %{} = payload) when is_channel(channel) do
+    notify(conf, channel, [payload])
+  end
+
+  def notify(%Config{} = conf, channel, [_ | _] = payload) when is_channel(channel) do
+    channel = "#{conf.prefix}.#{@mappings[channel]}"
+
+    Repo.query(
+      conf,
+      "SELECT pg_notify($1, payload) FROM json_array_elements_text($2::json) AS payload",
+      [channel, Enum.map(payload, &Jason.encode!/1)]
+    )
+
+    :ok
+  end
```

Comment:
For the sake of consistency it would be nice to make this `notify(server, channel, payload)`. Other implementations may rely on server state. We can fetch the `conf` from the local server state with something fast like this:

```elixir
with %State{conf: conf} <- GenServer.call(server, :get_state) do
  # use the conf in here
end
```

That pulls the server state into the local process and avoids bottlenecking outgoing notifications.

Reply:
Do you think we should do this for every function in the notifier? Seems like a more consistent approach. I think it would also be 100% compatible with the old implementation.